Lib/pyclbr.py

```diff
@@ -179,187 +169,113 @@ def _readmodule(module, path, inpackage=None):
     return _create_tree(fullmodule, path, fname, source, tree, inpackage)
 
 
-def _create_tree(fullmodule, path, fname, source, tree, inpackage):
-    """Return the tree for a particular module.
-
-    fullmodule (full module name), inpackage+module, becomes o.module.
-    path is passed to recursive calls of _readmodule.
-    fname becomes o.file.
-    source is tokenized.  Imports cause recursive calls to _readmodule.
-    tree is {} or {'__path__': <submodule search locations>}.
-    inpackage, None or string, is passed to recursive calls of _readmodule.
-
-    The effect of recursive calls is mutation of global _modules.
-    """
-    f = io.StringIO(source)
+class _ClassBrowser(ast.NodeVisitor):
+    def __init__(self, module, path, file, tree, inpackage):
+        self.path = path
+        self.tree = tree
+        self.file = file
+        self.module = module
+        self.inpackage = inpackage
+        self.stack = []
+
+    def visit_ClassDef(self, node):
+        bases = []
+        for base in node.bases:
+            name = ast.unparse(base)
+            if name in self.tree:
+                # We know this super class.
+                bases.append(self.tree[name])
+            elif len(names := name.split(".")) > 1:
+                # Super class form is module.class:
+                # look in module for class.
+                module = names[-2]
+                class_ = names[-1]
+                if module in _modules:
+                    bases.append(_modules[module].get(class_) or name)
+            else:
+                bases.append(name)
+
+        parent = self.stack[-1] if self.stack else None
+        class_ = Class(
+            self.module, node.name, bases, self.file, node.lineno, parent
+        )
+        if parent is None:
+            self.tree[node.name] = class_
+        self.stack.append(class_)
+        self.generic_visit(node)
+        self.stack.pop()
+
+    def visit_Assign(self, node):
+        if (
+            not len(node.targets) == 1
+            or not len(self.stack) > 0
+            or not isinstance(self.stack[-1], Class)
+            or not isinstance(node.targets[0], ast.Name)
+            or not isinstance(node.value, ast.Name)
+            or not isinstance(self.tree.get(node.value.id), Function)
+        ):
+            return
+
+        name = node.targets[0].id
+        child = copy.deepcopy(self.tree[node.value.id])
```

Comment:
Hummm.....this can be very inefficient as we would be copying the while tree for this id, no?

Reply:
we would only copy a `Function` object, which has really small memory footprint. If `Function` would have many children it would be inefficient but I think it is uncommon to put more than like 5-6 functions inside of another function.

Reply:
I agree, but although minor I wish we could avoid the copy....

Reply:
Ah, I think I can avoid this.

Reply:
But we always need to deepcopy children I guess :/

Reply:
```py
        name = node.targets[0].id
        value = self.tree[node.value.id]
        parent = self.stack[-1]
        child = Function(
            value.module, name, value.file, node.lineno, parent, value.is_async
        )
        child.children = copy.deepcopy(value.children)
        parent._addchild(name, child)
        parent._addmethod(name, node.lineno)
``` what do you think about this?